src/core/mem_throt.c

```diff
@@ -5,63 +5,122 @@
+//         }
+//     }
 
 void mem_throt_period_timer_callback(irqid_t int_id) {
 
     timer_disable();
-    events_cntr_disable(cpu()->mem_throt.counter_id);
-    timer_reschedule_interrupt(cpu()->mem_throt.period_counts);
-    events_cntr_set(cpu()->mem_throt.counter_id, cpu()->mem_throt.budget);
-
-    if (cpu()->mem_throt.throttled) {
-        events_cntr_irq_enable(cpu()->mem_throt.counter_id);
-        cpu()->mem_throt.throttled = false;
+    events_cntr_disable(cpu()->vcpu->vm->mem_throt.counter_id);
+    timer_reschedule_interrupt(cpu()->vcpu->vm->mem_throt.period_counts);
+    events_cntr_set(cpu()->vcpu->vm->mem_throt.counter_id, cpu()->vcpu->vm->mem_throt.budget);
+
+    if (cpu()->vcpu->vm->mem_throt.throttled) 
+    {
+        events_cntr_irq_enable(cpu()->vcpu->vm->mem_throt.counter_id);
+        cpu()->vcpu->vm->mem_throt.throttled = false;
     }
-
-    events_cntr_enable(cpu()->mem_throt.counter_id);
+    events_cntr_enable(cpu()->vcpu->vm->mem_throt.counter_id);
+    
+    if (cpu()->vcpu->vm->master) 
+        cpu()->vcpu->vm->mem_throt.ticket_num_left = cpu()->vcpu->vm->mem_throt.ticket_num;
+    
     timer_enable();
-}
+    // uint32_t target_cpu_mask = cpu()->vcpu->vm->cpus;  // Assuming this is a bitmask of all CPUs
+    
+    // Clear the bit for the current CPU to avoid sending IPI to itself
+    // target_cpu_mask &= ~(1ull << cpu()->id);  // Exclude the current CPU from the mask
+    // Send IPI to each CPU in the target mask using SGI for standby
+    // for (int i = 0; i < sizeof(target_cpu_mask) * 8; i++) {
+    //     if (target_cpu_mask & (1ull << i)) {            
+    //         // Create the cpu_msg structure here
+    //         struct cpu_msg msg = {
+    //             .event = THROT_CPU_WAKEUP,
+    //             .handler = THROT_IPI_ID
+    //             };
+    //         console_printk("Sending IPI with event %lu, handler %d to cpu %d\n", msg.event, msg.handler, i);
+    //         cpu_send_msg(i, &msg);
+    //     }
+    //  }
 
+}
+void mem_throt_event_overflow_callback(irqid_t int_id) {
 
+    events_clear_cntr_ovs(cpu()->vcpu->vm->mem_throt.counter_id);
+    events_cntr_disable(cpu()->vcpu->vm->mem_throt.counter_id);
+    events_cntr_irq_disable(cpu()->vcpu->vm->mem_throt.counter_id);
 
-void mem_throt_event_overflow_callback(irqid_t int_id) {
+    if(cpu()->vcpu->vm->mem_throt.ticket_num_left > 0)
+    {
+        cpu()->vcpu->vm->mem_throt.ticket_num_left--;
+        events_cntr_set(cpu()->vcpu->vm->mem_throt.counter_id, cpu()->vcpu->vm->mem_throt.budget);
+        events_cntr_enable(cpu()->vcpu->vm->mem_throt.counter_id);
 
-    events_clear_cntr_ovs(cpu()->mem_throt.counter_id);
-    events_cntr_disable(cpu()->mem_throt.counter_id);
-    events_cntr_irq_disable(cpu()->mem_throt.counter_id);
-    cpu()->mem_throt.throttled = true;
-    cpu_standby();
+    }
+    else
+    {
+        cpu()->vcpu->vm->mem_throt.throttled = true;  
+        cpu_standby();
+    }
+    //enhace this part
```

Comment:
what should be done here?

Reply:
It's already done, I'll remove the comment.